src/server/tiering/external_alloc.cc

```diff
@@ -478,12 +496,23 @@ void ExternalAllocator::FreePage(Page* page, SegmentDescr* owner, size_t block_s
   BinIdx bidx = ToBinIdx(block_size);
 
   // Remove fast allocation reference.
-  if (free_pages_[bidx] == page) {
-    free_pages_[bidx] = &empty_page;
+  if (page->free_chained) {
+    if (free_pages_[bidx] == page) {
+      free_pages_[bidx] = &empty_page;
+    } else {
+      for (auto* cur = free_pages_[bidx]; cur != nullptr; cur = cur->next_free) {
```

Comment:
i do not mind. just a suggestion - another way to do it - is lazily, so when you actually pull a full page from the `free_pages_` list, you pop all the "empty" ones, though one can argue it's a premature optimisation :)

Reply:
Also a very fragile approach, we return the page to the segment and might decide to use with a different bin index. We can't keep the pointer